Improved how Wasp prints output of underlying processes.

Martinsos opened the pull request:
Fixes #793

Martinsos commented on waspc/cli/src/Wasp/Cli/Command.hs:
```diff
@@ -17,8 +17,7 @@ newtype Command a = Command {_runCommand :: ExceptT CommandError IO a}
 runCommand cmd = do
-  errorOrResult <- runExceptT $ _runCommand cmd
-  case errorOrResult of
+  runExceptT (_runCommand cmd) >>= \case
```
Unrelated refactoring, but it is nice so I did it!

It is cool, with `\case` (which comes from LambdaCase extension) we can directly feed the result of monadic action into the case instead of having to extract it into a variable first, nice!
@matijaSos this is what I mentioned today to you.

Martinsos commented on waspc/src/Wasp/Generator/Job/Common.hs:
```diff
@@ -0,0 +1,24 @@
+module Wasp.Generator.Job.Common
```
I just moved this stuff from another file here, this is not new.

Martinsos commented on waspc/src/Wasp/Generator/NpmInstall.hs:
```diff
@@ -101,25 +102,27 @@ loadInstalledFullStackNpmDependencies dstDir = do
-          jobMsg <- readChan chan
+        go :: (Bool, Bool) -> Chan J.JobMessage -> PrefixedWriter ()
+        go (True, True) _ = return ()
+        go (isWebAppDone, isServerDone) chan = do
+          jobMsg <- liftIO $ readChan chan
           case J._data jobMsg of
             J.JobOutput {} ->
-              printPrefixedJobMessage prevJobMsg jobMsg
-                >> go (Just jobMsg) (isWebAppDone, isServerDone) chan
+              printPrefixedJobMessage jobMsg
+                >> go (isWebAppDone, isServerDone) chan
             J.JobExit {} -> case J._jobType jobMsg of
-              J.WebApp -> go (Just jobMsg) (True, isServerDone) chan
-              J.Server -> go (Just jobMsg) (isWebAppDone, True) chan
+              J.WebApp -> go (True, isServerDone) chan
+              J.Server -> go (isWebAppDone, True) chan
```
Introducing state monad allowed us to hide explicit state from here!
So now here we can call prefixed writing without having to know about how it keeps its state! It just works. Before state was "previous message", now it is not visible. Super cool!

shayneczyzewski replied:
That is very cool you abstracted away the previous message - nice! 👍🏻

shayneczyzewski commented on waspc/src/Wasp/Generator/Job/Common.hs:
```diff
@@ -0,0 +1,24 @@
+import System.IO (Handle, stderr, stdout)
+import qualified Wasp.Generator.Job as J
+
+getJobMessageOutHandle :: J.JobMessage -> Handle
```
Tricky call. I think both are valid. I would say here slightly since they are more specific utility than just operating on the types themselves as generic helpers.

shayneczyzewski commented on waspc/src/Wasp/Generator/Job/IO/PrefixedWriter.hs:
```diff
@@ -0,0 +1,125 @@
+    addPrefixWhereNeeded =
+      let prefix = makeJobMessagePrefix jobMessage
+          addPrefixAfterSubstr substr = T.intercalate (substr <> prefix) . T.splitOn substr
+       in addPrefixAfterSubstr "\r" . addPrefixAfterSubstr "\n"
```
Why again are we splitting the `\r` and `\n` cases like this again, vs treating them as `\r\n`?

Martinsos replied:
Pfuf, those are good questions!
`\r` is valid because it can be used to "reset the cartridge" -> so it moves cursor to the start of the line removing previous output and writes what comes after it. It can be used for stuff like showing a progress bar, or percentages growing. Prisma uses it in some of its output, I think npm also possibly (e.g. when installing stuff).

As for windows -> I know they use \r\n in the files, but maybe they don't in the shell?
I would maybe just ignore this one for now, and we can handle that when the windows come :D.

shayneczyzewski replied:
Ah, yes, silly Prisma-like executables. Forgot they do carriage return stuff and some other funky ANSI symbols. Ok, makes sense! Yes, let's punt on Windows for now! :D

Martinsos replied:
Hey `\r` is not that weird, we are also going to use it for `npm install` progress indicator probably :D!
Btw I am going to leave a TODO here to think more about Windows in the future.

shayneczyzewski replied:
Haha, yeah it's true `\r` is fine for interactive outputs, I was just joking since I spaced on it. :D TODO for Windows makes sense. 👍🏻

shayneczyzewski commented on waspc/src/Wasp/Generator/Job/IO/PrefixedWriter.hs:
```diff
@@ -0,0 +1,125 @@
+      if contentEndsWithNewline
+        then S.insert jobType jobsWithPendingNewline
+        else S.delete jobType jobsWithPendingNewline
+    contentEndsWithNewline = "\n" `T.isSuffixOf` content
```
I found some of the inner "if-ing" a bit hard to grok... is this any easier perhaps?

```haskell
applyPendingNewline ::
  JobsWithPendingNewline -> (J.JobType, T.Text) -> (JobsWithPendingNewline, T.Text)
applyPendingNewline jobsWithPendingNewline (jobType, content) =
  let content' = addPendingNewlineToStartIfAny $ removeTrailingNewlineIfAny content
      jobsWithPendingNewlines' = updateOp jobType jobsWithPendingNewline
   in (jobsWithPendingNewlines', content')
  where
    updateOp :: JobType -> Set JobType -> Set JobType
    updateOp = if contentEndsWithNewline then S.insert else S.delete

    addPendingNewlineToStartIfAny = if jobType `S.member` jobsWithPendingNewline then ("\n" <>) else id
    removeTrailingNewlineIfAny = if contentEndsWithNewline then T.init else id
    contentEndsWithNewline = "\n" `T.isSuffixOf` content
```

Martinsos replied:
I agree this part is a bit complex, I tried simplifying it couple of times but never got to perfect solution. This looks interesting I will try using it, maybe doing some hybrid!

Martinsos replied:
I went with something in between and ended up with this:
```hs
applyPendingNewline ::
  JobsWithPendingNewline -> (J.JobType, T.Text) -> (JobsWithPendingNewline, T.Text)
applyPendingNewline jobsWithPendingNewline (jobType, content) = (jobsWithPendingNewline', content')
  where
    content' = addPendingNewlineToStartIfAny $ removeTrailingNewlineIfAny content
      where
        removeTrailingNewlineIfAny =
          if contentEndsWithNewline then T.init else id
        addPendingNewlineToStartIfAny =
          if jobType `S.member` jobsWithPendingNewline then ("\n" <>) else id
    jobsWithPendingNewline' = updateOp jobType jobsWithPendingNewline
      where
        updateOp = if contentEndsWithNewline then S.insert else S.delete
    contentEndsWithNewline = "\n" `T.isSuffixOf` content
```

What does it look like?

shayneczyzewski replied:
Looks good to me!

shayneczyzewski reviewed (commented):
Hey @Martinsos - this is super cool! It made my head hurt, so I am taking a pause for now and will re-review tomorrow with fresh eyes/brain :D Dang Monad Transformers and all that stuff haha

One quick question, is there a clean way to write tests for this? I just did some playing in the repl, and may have found an undesired case, but I wonder if we can do some sort of fuzzing/prop testing (or even explicit unit tests) around this? Maybe not worth it?

Anyways, when I try the following:
```haskell
:set +m

import Wasp.Generator.Job.IO.PrefixedWriter (PrefixedWriter, printPrefixedJobMessage, runPrefixedWriter)
import qualified Wasp.Generator.Job as J

runPrefixedWriter $ do
  printPrefixedJobMessage $ J.JobMessage { J._data = J.JobOutput "line 1a" J.Stdout, J._jobType = J.Server }
  printPrefixedJobMessage $ J.JobMessage { J._data = J.JobOutput "one\n" J.Stdout, J._jobType = J.Db }
  printPrefixedJobMessage $ J.JobMessage { J._data = J.JobOutput "line 1b\n" J.Stdout, J._jobType = J.Server }
  printPrefixedJobMessage $ J.JobMessage { J._data = J.JobOutput "line 2\n" J.Stdout, J._jobType = J.Server }
```

I get this interleaving:
```bash
Server: line 1a
Db: oneline 1b
Server: line 2
```

I am guessing this second output line is not a case we want, or are we explicitly not protecting against that? Anyways, great work, I will resume tomorrow after the brain break!

Martinsos commented:
> Hey @Martinsos - this is super cool! It made my head hurt, so I am taking a pause for now and will re-review tomorrow with fresh eyes/brain :D Dang Monad Transformers and all that stuff haha
> 
> One quick question, is there a clean way to write tests for this? I just did some playing in the repl, and may have found an undesired case, but I wonder if we can do some sort of fuzzing/prop testing (or even explicit unit tests) around this? Maybe not worth it?
> 
> Anyways, when I try the following:
> 
> ```haskell
> :set +m
> 
> import Wasp.Generator.Job.IO.PrefixedWriter (PrefixedWriter, printPrefixedJobMessage, runPrefixedWriter)
> import qualified Wasp.Generator.Job as J
> 
> runPrefixedWriter $ do
>   printPrefixedJobMessage $ J.JobMessage { J._data = J.JobOutput "line 1a" J.Stdout, J._jobType = J.Server }
>   printPrefixedJobMessage $ J.JobMessage { J._data = J.JobOutput "one\n" J.Stdout, J._jobType = J.Db }
>   printPrefixedJobMessage $ J.JobMessage { J._data = J.JobOutput "line 1b\n" J.Stdout, J._jobType = J.Server }
>   printPrefixedJobMessage $ J.JobMessage { J._data = J.JobOutput "line 2\n" J.Stdout, J._jobType = J.Server }
> ```
> 
> I get this interleaving:
> 
> ```shell
> Server: line 1a
> Db: oneline 1b
> Server: line 2
> ```
> 
> I am guessing this second output line is not a case we want, or are we explicitly not protecting against that? Anyways, great work, I will resume tomorrow after the brain break!

You are right, this is problematic case! I on purpose gave up on handling it for now, but now that you write about it, I realized we did have that solved before, since we had information about the previous message. We could add that information back to the state of the PrefixedWriter and again handle it.
The thing is, we can't really defend ourselves completely -> there can be output from non-job, and we will have no idea about it, and we will get this issue, we can't avoid it.

Good thing though is, that in practice, messages usually don't came broken into pieces like that, mostly they come with newline at the end. Even if they do come broken in the middle, which is pretty / super rare from what I saw, they will then very likely come very quickly after each other, again making the chance of this happening very low. So I would say in practice, this is not likely to happen at all.

I think it still might be worth tracking the last message and defending against this use case, I will add that.

shayneczyzewski commented:
> > Hey @Martinsos - this is super cool! It made my head hurt, so I am taking a pause for now and will re-review tomorrow with fresh eyes/brain :D Dang Monad Transformers and all that stuff haha
> > One quick question, is there a clean way to write tests for this? I just did some playing in the repl, and may have found an undesired case, but I wonder if we can do some sort of fuzzing/prop testing (or even explicit unit tests) around this? Maybe not worth it?
> > Anyways, when I try the following:
> > ```haskell
> > :set +m
> > 
> > import Wasp.Generator.Job.IO.PrefixedWriter (PrefixedWriter, printPrefixedJobMessage, runPrefixedWriter)
> > import qualified Wasp.Generator.Job as J
> > 
> > runPrefixedWriter $ do
> >   printPrefixedJobMessage $ J.JobMessage { J._data = J.JobOutput "line 1a" J.Stdout, J._jobType = J.Server }
> >   printPrefixedJobMessage $ J.JobMessage { J._data = J.JobOutput "one\n" J.Stdout, J._jobType = J.Db }
> >   printPrefixedJobMessage $ J.JobMessage { J._data = J.JobOutput "line 1b\n" J.Stdout, J._jobType = J.Server }
> >   printPrefixedJobMessage $ J.JobMessage { J._data = J.JobOutput "line 2\n" J.Stdout, J._jobType = J.Server }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I get this interleaving:
> > ```shell
> > Server: line 1a
> > Db: oneline 1b
> > Server: line 2
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I am guessing this second output line is not a case we want, or are we explicitly not protecting against that? Anyways, great work, I will resume tomorrow after the brain break!
> 
> You are right, this is problematic case! I on purpose gave up on handling it for now, but now that you write about it, I realized we did have that solved before, since we had information about the previous message. We could add that information back to the state of the PrefixedWriter and again handle it. The thing is, we can't really defend ourselves completely -> there can be output from non-job, and we will have no idea about it, and we will get this issue, we can't avoid it.
> 
> Good thing though is, that in practice, messages usually don't came broken into pieces like that, mostly they come with newline at the end. Even if they do come broken in the middle, which is pretty / super rare from what I saw, they will then very likely come very quickly after each other, again making the chance of this happening very low. So I would say in practice, this is not likely to happen at all.
> 
> I think it still might be worth tracking the last message and defending against this use case, I will add that.

I think you are right, it feels like a very unlikely scenario. I don't think we should invest much time on it now (possibly ever), but whether we want to create an issue or not is your call. I'm good either way, just wanted to note it and confirm! 👍🏻 This is really looking nice!

Martinsos commented:
> > > Hey @Martinsos - this is super cool! It made my head hurt, so I am taking a pause for now and will re-review tomorrow with fresh eyes/brain :D Dang Monad Transformers and all that stuff haha
> > > One quick question, is there a clean way to write tests for this? I just did some playing in the repl, and may have found an undesired case, but I wonder if we can do some sort of fuzzing/prop testing (or even explicit unit tests) around this? Maybe not worth it?
> > > Anyways, when I try the following:
> > > ```haskell
> > > :set +m
> > > 
> > > import Wasp.Generator.Job.IO.PrefixedWriter (PrefixedWriter, printPrefixedJobMessage, runPrefixedWriter)
> > > import qualified Wasp.Generator.Job as J
> > > 
> > > runPrefixedWriter $ do
> > >   printPrefixedJobMessage $ J.JobMessage { J._data = J.JobOutput "line 1a" J.Stdout, J._jobType = J.Server }
> > >   printPrefixedJobMessage $ J.JobMessage { J._data = J.JobOutput "one\n" J.Stdout, J._jobType = J.Db }
> > >   printPrefixedJobMessage $ J.JobMessage { J._data = J.JobOutput "line 1b\n" J.Stdout, J._jobType = J.Server }
> > >   printPrefixedJobMessage $ J.JobMessage { J._data = J.JobOutput "line 2\n" J.Stdout, J._jobType = J.Server }
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > I get this interleaving:
> > > ```shell
> > > Server: line 1a
> > > Db: oneline 1b
> > > Server: line 2
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > I am guessing this second output line is not a case we want, or are we explicitly not protecting against that? Anyways, great work, I will resume tomorrow after the brain break!
> > 
> > 
> > You are right, this is problematic case! I on purpose gave up on handling it for now, but now that you write about it, I realized we did have that solved before, since we had information about the previous message. We could add that information back to the state of the PrefixedWriter and again handle it. The thing is, we can't really defend ourselves completely -> there can be output from non-job, and we will have no idea about it, and we will get this issue, we can't avoid it.
> > Good thing though is, that in practice, messages usually don't came broken into pieces like that, mostly they come with newline at the end. Even if they do come broken in the middle, which is pretty / super rare from what I saw, they will then very likely come very quickly after each other, again making the chance of this happening very low. So I would say in practice, this is not likely to happen at all.
> > I think it still might be worth tracking the last message and defending against this use case, I will add that.
> 
> I think you are right, it feels like a very unlikely scenario. I don't think we should invest much time on it now (possibly ever), but whether we want to create an issue or not is your call. I'm good either way, just wanted to note it and confirm! 👍🏻 This is really looking nice!

@shayneczyzewski I also added this now to the prefixing logic, and while doing it, I also found another case with a missing newline sometimes -> anyway, this should be covered well now! I think it is good we added it, I feel the output is now more robust. It was also easy to add additional tracking of last message, which was cool.

While at it I also improved the looks of output a bit: gave nice color to "DB" and colored "stderr" in yellow to be more visible.

Here is a screenshot of what it look like now:
![image](https://user-images.githubusercontent.com/1536647/199559745-2eefc9b1-cb8e-48f8-a931-b3739991bff8.png)

Martinsos commented:
Well I just realized from the screenshot above that I completely forgot that some messages can come from stderr while some can come from stdout! And those we also don't want to mix!

I now modified code to also take that into account. So now it doesn't care just about job type, but about something called "output".

Uff, I think this should be it! Maybe take a quick look at the `Output` type @shayneczyzewski and tell me if that seems sound.

shayneczyzewski reviewed (approved):
Hey @Martinsos nice updates! Good catch on tracking `stdout` in the decision-making for prefixes. I think adding that bool to `Output` makes sense.

Glad to see tracking the last job message was made easier with `PrefixedWriterState` as well.

`printPrefixedJobMessage` does still feel a bit hard to follow to me sometimes, but I think you minimized as much complexity as you can and the detailed comment above helps readers as well. It is just a hard problem to solve, and the helpers do aid in understanding what is going on, so I think much of it is essential complexity.

I did a quick test and what is here seems to function nicely. 👍🏻 The only thing I would consider is possibly adding a space between the `JobType` and "(stderr)", so you would go from "Db(stderr):" to "Db (stderr):", but that may just be my visual preference.

Martinsos commented:
@shayneczyzewski you are right that printPrefixedJobMessage is still a bit too complex.
I now refactored it a bit more, take a look, I believe it is a bit more readable.
I also changed the name to printJobMessagePrefixed, since that is more correct -> we are not printing already prefixed job message, we prefix it ourselves and then print it.

As for space between Server and (stderr) -> there was actually space there before, but I removed it because it felt a bit like it was part of the input to the right. At least I subjectively thought so :D. So maybe let's leave it like this for now and we can change in the future if we figure out it is not the best?

shayneczyzewski commented:
> @shayneczyzewski you are right that printPrefixedJobMessage is still a bit too complex. I now refactored it a bit more, take a look, I believe it is a bit more readable. I also changed the name to printJobMessagePrefixed, since that is more correct -> we are not printing already prefixed job message, we prefix it ourselves and then print it.
> 
> I am sure it could be improved further, but it feels like I reached the threshold where refactoring pays off, I feel going further the ratio of invested time and what I got would not be the best.
> 
> As for space between Server and (stderr) -> there was actually space there before, but I removed it because it felt a bit like it was part of the input to the right. At least I subjectively thought so :D. So maybe let's leave it like this for now and we can change in the future if we figure out it is not the best?

Looking great! I like the refactor to simplify things a bit more. I think this will work just fine. 👍🏻 

No sweat on space, we can leave it as-is and see what others think. All good by me. 🚀
